### DISCUSSION TITLE
Problems with ActivatedRoute in Standalone Component

### DISCUSSION THREAD

**Discussion by Simon-TheHelpfulCat** [help-me-mom/ng-mocks#3634] [Q&A] [answered]
I am having an issue with using ```MockBuilder``` to test a Standalone Component that uses ```[RouterLink]```.

The Component is defined like this:
```Typescript
@Component({
  standalone: true,
  imports: [CommonModule, RouterModule, MatListModule],
  templateUrl: './my.component.html',
  styleUrls: ['./my.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class MyComponent {}
```
The template is:
```HTML
<mat-nav-list>
  <a mat-list-item [routerLink]="['link']">Link</a>
</mat-nav-list>
```
My test is still at the ```should create``` stage:
```Typescript
  it('should create', () => {
    const fixture = TestBed.createComponent(MyComponent);
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });
```
I have recently switched to using ```MockBuilder``` for my test setup and for this test I started with:
```Typescript
    MockBuilder(MyComponent);
```
but this resulted in:
> NullInjectorError: No provider for ActivatedRoute!

Looking through the documentation suggested that:
```Typescript
    MockBuilder(MyComponent)
      .keep(RouterModule)
      .keep(RouterTestingModule)
```
might work, but I get the same result. I also tried:
```Typescript
    MockBuilder(MyComponent)
      .mock(RouterModule)
      .keep(RouterTestingModule)
```
and
```Typescript
    MockBuilder(MyComponent)
      .exclude(RouterModule)
      .keep(RouterTestingModule)
```
with no luck. I even tried:
```Typescript
    MockBuilder(MyComponent)
      .provide(MockProvider(ActivatedRoute));
```
but it still tells me I have no provider for ```ActivatedRoute```. If I use ```TestBed``` directly:
```Typescript
   TestBed.configureTestingModule({
      imports: [MyComponent, RouterTestingModule]
    }).compileComponents();
```
the test passes.

It's probably something really stupid, but I just can't see it. Any help would be gratefully received.

**Comment by satanTime** [comment, marked as answer]
Hi there,

`ActivatedRoute` isn't a part of `RouterModule`, it's defined in `.forRoot` and in `.forChild`.
Because your component imports `RouterModule` only, you need to provide `ActivatedRoute` additionally for `TestBed`.

Your last try with `.provide` is the right one and works correctly: https://codesandbox.io/s/sweet-rubin-mx7owq?file=/src/test.spec.ts

**Reply by satanTime** [reply]
I found a couple of bugs, maybe one of them is causing issues on your side. If you can fail the example above like your project is failing, that would be great.

**Reply by Simon-TheHelpfulCat** [reply]
Okay, coming back to it this morning with fresh eyes I have spotted the problem. In actual fact:
```Typescript
MockBuilder(MyComponent)
```
is all that was needed

The issue was that I was not returning the builder. I had:
```Typescript
  beforeEach(async () => {
    MockBuilder(MyComponent)
  });
```
whereas I should have had:
```Typescript
  beforeEach(async () => MockBuilder(MyComponent));
```
or
```Typescript
  beforeEach(async () => {
    return MockBuilder(MyComponent)
  });
```
If I modify your example to make it how I originally had it, it fails in the same way.

I am sorry to have bothered you with such a trivial mistake, but I'm afraid that last night my tired eyes just couldn't see the error! Hopefully it will benefit another tired developer at some point by making them look again at their code.

**Reply by satanTime** [reply]
Hi Simon, good that it was a simple fix :) And no worries, feel free to ask a question any time.
